### PR TITLE
FP8 support in core module

### PR DESCRIPTION
**Pull request by abhishek-gola**
Core part of https://github.com/opencv/opencv/issues/29313

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
Looks like you have to hack CUDA and OpenCL paths too:
```
[ RUN      ] GpuMat.convertTo/13, where GetParam() = (8UC1, 8F_E4M3FNC1)
Segmentation fault (core dumped)
```

**Comment by asmorkalov** [comment]
RISC-V RVV:
```
[ RUN      ] Core_ConvertScaleAbs/ElemWiseTest.accuracy/0, where GetParam() = 16-byte object <D0-15 1A-56 55-55 00-00 90-13 1A-56 55-55 00-00>
/home/ci/opencv/modules/core/test/test_arithm.cpp:1780: Failure
too big relative difference (255 > 1) between 8uC4 3-dim (1 x 1 x 14717) 'dst0' and 'dst' at [0;
 0;
 1].
- dst0:
[255].
- dst:
[  0].


src[0] ~ ?C4 3-dim (1 x 1 x 14717)
testCase #18

[  FAILED  ] Core_ConvertScaleAbs/ElemWiseTest.accuracy/0, where GetParam() = 16-byte object <D0-15 1A-56 55-55 00-00 90-13 1A-56 55-55 00-00> (339 ms)
```

File: modules/core/include/opencv2/core/hal/interface.h [hunk 1]
```diff
@@ -64,12 +64,17 @@ typedef int16_t cv_hal_bf16;
 #define CV_64U  10
 #define CV_64S  11
 #define CV_32U  12
-#define CV_DEPTH_CURR_MAX 13
+#define CV_8F_E4M3FN   13  // OCP/ONNX FLOAT8E4M3FN  (1-4-3, bias 7, no inf, max 448)
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/hal/interface.h:67]: please, introduce `CV_8F` alias for `CV_8F_E4M3FN`, because it seems to be the main 8-bit floating-point type.

File: modules/core/include/opencv2/core/hal/interface.h [hunk 2]
```diff
@@ -64,12 +64,17 @@ typedef int16_t cv_hal_bf16;
...
+#define CV_8F_E4M3FNUZ 14  // FLOAT8E4M3FNUZ         (1-4-3, bias 8, no inf, single NaN, no -0)
+#define CV_8F_E5M2     15  // FLOAT8E5M2             (1-5-2, bias 15, IEEE-like, inf/NaN)
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/hal/interface.h]: I really suggest to drop those 2 flavors for now, `CV_8F_E5M2` and `CV_8F_E5M2FNUZ`, because, to the best of my knowledge, they are not used for tensors in pre-trained models, they are only used inside DL training frameworks to store gradients, where dynamic range is significantly more important than accuracy.

Any extra type to the existing `n` types would mean extra `2*n` conversion kernels.

File: modules/core/include/opencv2/core/hal/interface.h [hunk 3]
```diff
@@ -152,6 +157,15 @@ typedef int16_t cv_hal_bf16;
 #define CV_16BFC4 CV_MAKETYPE(CV_16BF,4)
 #define CV_16BFC(n) CV_MAKETYPE(CV_16BF,(n))
 
+#define CV_8F_E4M3FNC1 CV_MAKETYPE(CV_8F_E4M3FN,1)
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/hal/interface.h:159]: please, introduce `CV_8FC*` macros.

File: modules/core/include/opencv2/core/cvdef.h [hunk 4]
```diff
@@ -979,6 +982,80 @@ class bfloat
     ushort w;
 };
 
+template<int ebits, int mbits, int bias, bool has_inf, bool fnuz>
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/cvdef.h]: that's too slow conversion to be practical.

please, spend a little more time on it and introduce dedicated classes for `fp8_t` (e4m3) and `fp8a_t` (e4m3fnuz) structures with dedicated optimized conversions, because 99% of the time people will use `float8_t` type, some of AMD/HIP users will use `float8a_t` e.g.:

```
struct fp8_t
{
      fp8_t() : b(0) {}
      explicit fp8_t(float f) {
            Cv32suf u;
            u.f = f;
            uint32_t fbits = u.u;
            const unsigned sign  = fbits >> 31;
            const unsigned abs_f = fbits & 0x7FFFFFFFu;
            const unsigned e32_raw = abs_f >> 23;

            const unsigned abs_r = abs_f + 0x00080000u;
            const unsigned e32   = abs_r >> 23;
            const unsigned m3    = (abs_r >> 20) & 7u;

            unsigned e4     = e32 - 120u;
            unsigned result = (e4 << 3) | m3;
            result = std::min(result, 0x7Eu);

            if (e32_raw < 121u) {
               const unsigned man   = abs_f & 0x007FFFFFu;
               const unsigned full  = (1u << 23) | man;
               const unsigned shift = 121u - e32_raw;
               const unsigned m3    = (full + (1u << (shift + 19u))) >> (shift + 20u);
               b = (uchar)((sign << 7) | ((e32_raw >= 118u) ? (m3 & 0x07u) : 0));
            } else {
               if (abs_f >= 0x7F800000u)
                  result = (abs_f == 0x7F800000u) ? 0x7Eu : 0x7Fu; // Inf→448, NaN→NaN
               b = (uchar)((sign << 7) | result);
            }
      }
 
      operator float() const {
         static constexpr ushort e4m3bf16_tab[] = {
           0, 15104, 15232, 15296, 15360, 15392, 15424, 15456, 15488, 15504, 15520, 15536,
           15552, 15568, 15584, 15600, 15616, 15632, 15648, 15664, 15680, 15696, 15712, 15728,
           15744, 15760, 15776, 15792, 15808, 15824, 15840, 15856, 15872, 15888, 15904, 15920,
           15936, 15952, 15968, 15984, 16000, 16016, 16032, 16048, 16064, 16080, 16096, 16112,
           16128, 16144, 16160, 16176, 16192, 16208, 16224, 16240, 16256, 16272, 16288, 16304,
           16320, 16336, 16352, 16368, 16384, 16400, 16416, 16432, 16448, 16464, 16480, 16496,
           16512, 16528, 16544, 16560, 16576, 16592, 16608, 16624, 16640, 16656, 16672, 16688,
           16704, 16720, 16736, 16752, 16768, 16784, 16800, 16816, 16832, 16848, 16864, 16880,
           16896, 16912, 16928, 16944, 16960, 16976, 16992, 17008, 17024, 17040, 17056, 17072,
           17088, 17104, 17120, 17136, 17152, 17168, 17184, 17200, 17216, 17232, 17248, 17264,
           17280, 17296, 17312, 17328, 17344, 17360, 17376, 32704
         };

         uchar sign = b & 0x80;
         uchar abs_b = b & 0x7f;
         return (e4m3bf16_tab[abs_b] << 16) | (sign << 24);
     }

     uchar b;
};
```

File: modules/core/include/opencv2/core/cvdef.h [hunk 5]
```diff
@@ -979,6 +982,80 @@ class bfloat
...
+        {
+            if (has_inf) { Cv32suf inf; inf.u = sign ? 0xFF800000u : 0x7F800000u; return man == 0 ? inf.f : qn.f; }
+            if (man == (1u << mbits) - 1) return qn.f;
+        }
+        if (exp == 0) return s * (float)man * pow2(1 - bias - mbits);               // subnormal
+        return s * (1.0f + (float)man / (float)(1u << mbits)) * pow2((int)exp - bias);
+    }
+
+protected:
+    uchar w;
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/cvdef.h]: use `uchar b`, 'b' for byte. We used `w` for hfloat/bfloat because there it was a 'word' (16-bit unsigned integer).

File: modules/core/include/opencv2/core/cvdef.h [hunk 6]
```diff
@@ -979,6 +982,80 @@ class bfloat
...
+        if (shift <= 0) return full << (-shift);
+        unsigned q = full >> shift, rem = full & ((1u << shift) - 1), half = 1u << (shift - 1);
+        if (rem > half || (rem == half && (q & 1))) q++;
+        return q;
+    }
+    static float pow2(int n) { Cv32suf s; s.u = (unsigned)((n + 127) << 23); return s.f; } // 2^n, n in [-126,127]
+};
+
+typedef float8_t<4, 3, 7,  false, false> float8_e4m3fn;   // no inf, max 448
+typedef float8_t<4, 3, 8,  false, true>  float8_e4m3fnuz; // no inf, single NaN, no -0
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/cvdef.h]: as mentioned above, `float8_e3m3fn` -> `fp8_t`, `float8_e4m3fnuz` -> `fp8a_t`, and we don't need e5 flavors

File: modules/core/src/check.cpp [hunk 7]
```diff
@@ -46,7 +46,8 @@ static const char* getTestOpMath(unsigned testOp)
 const char* depthToString_(int depth)
 {
     static const char* depthNames[] = { "CV_8U", "CV_8S", "CV_16U", "CV_16S", "CV_32S", "CV_32F", "CV_64F", "CV_16F",
-                                        "CV_16BF", "CV_Bool", "CV_64U", "CV_64S", "CV_32U" };
+                                        "CV_16BF", "CV_Bool", "CV_64U", "CV_64S", "CV_32U",
+                                        "CV_8F_E4M3FN", "CV_8F_E4M3FNUZ", "CV_8F_E5M2", "CV_8F_E5M2FNUZ" };
```
Comment (vpisarev) [inline modules/core/src/check.cpp]: please, use shorter names

File: modules/core/src/convert_scale.simd.hpp [hunk 8]
```diff
@@ -443,6 +465,43 @@ BinaryFunc getCvtScaleAbsFunc(int depth)
...
+#define DEF_CVT_SCALE_SCALAR_FUNC(suffix, stype, dtype, wtype) \
+static void cvtScale##suffix( const uchar* src_, size_t sstep, const uchar*, size_t, \
+                              uchar* dst_, size_t dstep, Size size, void* scale_) \
+{ \
+    const stype* src = (const stype*)src_; dtype* dst = (dtype*)dst_; \
+    const double* scale = (const double*)scale_; \
+    wtype a = (wtype)scale[0], b = (wtype)scale[1]; \
+    sstep /= sizeof(src[0]); dstep /= sizeof(dst[0]); \
+    for( int i = 0; i < size.height; i++, src += sstep, dst += dstep ) \
+        for (int j = 0; j < size.width; j++) \
```
Comment (vpisarev) [inline modules/core/src/convert_scale.simd.hpp:483]: let's do less, but with better quality. Please, implement optimized conversion from fp8_t to float using universal intrinsics.

File: modules/core/src/rand.cpp [hunk 9]
```diff
@@ -280,7 +292,9 @@ static RandFunc randTab[CV_DEPTH_MAX][CV_DEPTH_MAX] =
         (RandFunc)randi_16s, (RandFunc)randi_32s, (RandFunc)randf_16_or_32f,
         (RandFunc)randf_64f, (RandFunc)randf_16_or_32f, (RandFunc)randf_16_or_32f,
         (RandFunc)randi_8b, (RandFunc)randi_64u, (RandFunc)randi_64s,
-        (RandFunc)randi_32u, 0, 0, 0
+        (RandFunc)randi_32u,
+        (RandFunc)randf_16_or_32f, (RandFunc)randf_16_or_32f,   // CV_8F_E4M3FN, E4M3FNUZ
```
Comment (vpisarev) [inline modules/core/src/rand.cpp]: is it correct?

File: modules/core/test/test_fp8.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+#include "test_precomp.hpp"
+
+namespace opencv_test { namespace {
+
+// The four FP8 depths and their wrapper types share one set of expectations.
+// Values chosen to be exactly representable (so round-trips are bit-exact) plus
+// the special/overflow cases that distinguish the formats.
+
+TEST(Core_FP8, type_basics)
```
Comment (vpisarev) [inline modules/core/test/test_fp8.cpp:13]: please, also extend existing type conversion tests to include fp8 (both flavors).

File: modules/core/include/opencv2/core/traits.hpp [hunk 11]
```diff
@@ -336,6 +336,50 @@ template<> class DataType<bfloat>
          };
 };
 
+template<> class DataType<float8_e4m3fn>
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/traits.hpp]: please, extend FileStorage to read and write fp8 values (scalars and arrays of fp8 values); add the respective tests

**Review by vpisarev — CHANGES_REQUESTED** [review]
very good start! but, please, look at the comment and address them.

**Comment by asmorkalov** [comment]
OpenCV core issue on RISC-V (128-bit RVV):
```
[ RUN      ] Core_ConvertScaleAbs/ElemWiseTest.accuracy/0, where GetParam() = 16-byte object <00-CB 17-56 55-55 00-00 C0-C8 17-56 55-55 00-00>
/home/ci/opencv/modules/core/test/test_arithm.cpp:1780: Failure
too big relative difference (255 > 1) between 8uC3 2-dim (2488 x 2) 'dst0' and 'dst' at [0;
 0].
- dst0:
*255*, 255;
255, 255.
- dst:
*0*, 0;
0, 0.


src[0] ~ ?C3 2-dim (2488 x 2)
testCase #23

[  FAILED  ] Core_ConvertScaleAbs/ElemWiseTest.accuracy/0, where GetParam() = 16-byte object <00-CB 17-56 55-55 00-00 C0-C8 17-56 55-55 00-00> (270 ms)
```

File: modules/core/src/convert_scale.simd.hpp [hunk 12]
```diff
@@ -22,7 +22,8 @@ template<typename _Ts, typename _Td> inline void
 cvtabs_32f( const _Ts* src, size_t sstep, _Td* dst, size_t dstep,
             Size size, float a, float b )
 {
-#if (CV_SIMD || CV_SIMD_SCALABLE)
+// Excluding GNU in CV_SIMD_SCALABLE because of "opencv/issues/26936"
+#if (CV_SIMD || (CV_SIMD_SCALABLE && !(defined(__GNUC__) && !defined(__clang__))) )
```
Comment (asmorkalov) [inline modules/core/src/convert_scale.simd.hpp:26]: 1. There are no compilers besides GCC and Clang we support for SCALABLE branch. So the condition is ambiguous.
2. Affects all conversions on RISC-V RVV.

**Comment by asmorkalov** [comment]
I just re-run tests om RISC-V RVV device (Spacemit K1, RVV 256 bit) and do not see the test failure even without the last fix commit. It means, that the bug correlates with register size.

**Comment by abhishek-gola** [comment]
> I just re-run tests om RISC-V RVV device (Spacemit K1, RVV 256 bit) and do not see the test failure even without the last fix commit. It means, that the bug correlates with register size.

Got it, let me check once again

File: modules/core/src/convert_scale.simd.hpp [hunk 13]
```diff
@@ -32,7 +33,8 @@ cvtabs_32f( const _Ts* src, size_t sstep, _Td* dst, size_t dstep,
     for( int i = 0; i < size.height; i++, src += sstep, dst += dstep )
     {
         int j = 0;
-#if (CV_SIMD || CV_SIMD_SCALABLE)
+// Excluding GNU in CV_SIMD_SCALABLE because of "opencv/issues/26936"
+#if (CV_SIMD || (CV_SIMD_SCALABLE && !(defined(__GNUC__) && !defined(__clang__))) )
```
Comment (asmorkalov) [inline modules/core/src/convert_scale.simd.hpp]: Similar remark.

**Comment by asmorkalov** [comment]
@abhishek-gola What is the PR status and next steps here?

**Comment by abhishek-gola** [comment]
> @abhishek-gola What is the PR status and next steps here?

It's finalized from my end.

**Review by vpisarev — APPROVED** [review]
